worlds/pokemon_crystal/client.py

```diff
@@ -0,0 +1,258 @@
+
+            rom_name = bytes([byte for byte in rom_info[0] if byte != 0]).decode("ascii")
+            rom_version = int.from_bytes(rom_info[1], "little")
+
+            if rom_name == "PM_CRYSTAL":
+                logger.info("ERROR: You appear to be running an unpatched version of Pokemon Crystal. "
+                            "You need to generate a patch file and use it to create a patched ROM.")
+                return False
+            if rom_name != "AP_CRYSTAL":
+                return False
+            if rom_version != data.rom_version:
+                generator_version = "{0:x}".format(rom_version)
+                client_version = "{0:x}".format(data.rom_version)
+                logger.info("ERROR: The patch file used to create this ROM is not compatible with "
+                            "this client. Double check your version of pokemon_crystal.apworld "
+                            "against the version used to generate this game.")
+                logger.info(f"Client checksum: {client_version}, Generator checksum: {generator_version}")
+                return False
+        except UnicodeDecodeError:
+            return False
+        except bizhawk.RequestFailedError:
+            return False  # Should verify on the next pass
+
+        ctx.game = self.game
+        ctx.items_handling = 0b001
+        ctx.want_slot_data = True
+        ctx.watcher_timeout = 0.125
+        return True
+
+    async def set_auth(self, ctx: "BizHawkClientContext") -> None:
+        slot_name_bytes = (await bizhawk.read(ctx.bizhawk_ctx, [(data.rom_addresses["AP_Seed_Name"], 64, "ROM")]))[0]
+        print(bytes([byte for byte in slot_name_bytes if byte != 0]).decode("utf-8"))
+        ctx.auth = bytes([byte for byte in slot_name_bytes if byte != 0]).decode("utf-8")
+
+    async def game_watcher(self, ctx: "BizHawkClientContext") -> None:
+        if ctx.slot_data is not None:
+            if ctx.slot_data["goal"] == 0:
+                self.goal_flag = data.event_flags["EVENT_BEAT_ELITE_FOUR"]
+            else:
+                self.goal_flag = data.event_flags["EVENT_BEAT_RED"]
+        try:
+            overworld_guard = (data.ram_addresses["wArchipelagoSafeWrite"], [1], "WRAM")
+
+            read_result = await bizhawk.guarded_read(
+                ctx.bizhawk_ctx, [(data.ram_addresses["wArchipelagoItemReceived"], 5, "WRAM")], [overworld_guard])
+
+            if read_result is None:  # Not in overworld
+                return
+
+            num_received_items = int.from_bytes([read_result[0][1], read_result[0][2]], "little")
+            received_item_is_empty = read_result[0][0] == 0
+            phone_trap_index = read_result[0][4]
+
+            if num_received_items < len(ctx.items_received) and received_item_is_empty:
+                next_item = ctx.items_received[num_received_items].item
+                next_item = next_item if next_item < 256 else next_item - 256
```

Comment:
This line could use a quick comment for what this is for. Is it TM item names maybe?